File: rogue-server/check/checks/speed/speed_a.py
import math
from abc import ABC

from check.movement_check import MovementCheck
from util.potion_effect_type import PotionEffectType


class SpeedA(MovementCheck, ABC):

    def __init__(self, data):
        super().__init__(data)

        self.last_friction = .91
        self.last_offset_x = 0.026

    def handle(self, event):
        offset_x = event['distanceX']
        offset_y = event['distanceY']

        movement_speed = self.data.movement_tracker.walk_speed / 2
        jump_height = 0.42 + self.data.potion_tracker.get_potion_level(PotionEffectType.JUMP_BOOST) * 0.1

        if self.data.collision_tracker.previous_collisions['onGround']:
            # We always want to assume the player is sprinting, as
            # tracking the player's sprinting state via action packets
            # will not work in the event that the player's sprinting state desyncs

            movement_speed *= 1.3
            self.last_friction *= 0.91
            movement_speed *= 0.16277136 / math.pow(self.last_friction, 3)

            if self.data.collision_tracker.previous_collisions['underBlock'] or 0.001 < offset_y < jump_height:
                movement_speed += 0.2
        else:
            movement_speed = 0.026
            self.last_friction = 0.91

        movement_speed += self.data.movement_tracker.get_velocity_horizontal()
        movement_speed += (0.2 * self.data.potion_tracker.get_potion_level(PotionEffectType.SPEED))
        movement_speed -= (0.15 * self.data.potion_tracker.get_potion_level(PotionEffectType.SLOWNESS))

        ratio = (offset_x - self.last_offset_x) / movement_speed
        
        if ratio > 1.0 and not self.data.movement_tracker.teleporting:
            if offset_x > 0.2 and self.increment_buffer(1) > 3 and not self.data.movement_tracker.teleporting:
                self.fail('ratio', ratio, 'offset_x', offset_x, 'last', self.last_offset_x,
                          'movement_speed', movement_speed)
        else:
            self.decrement_buffer(0.25)

        self.last_offset_x = offset_x * self.last_friction
        self.last_friction = self.data.collision_tracker.previous_collisions['frictionFactor']

    def get_check_name(self):
        return 'Speed'

    def get_check_type(self):
        return 'A'

File: rogue-server/src/main/java/org/hostile/anticheat/check/impl/speed/SpeedA.java
package org.hostile.anticheat.check.impl.speed;

import org.hostile.anticheat.check.annotation.CheckMetadata;
import org.hostile.anticheat.check.type.impl.PositionUpdateCheck;
import org.hostile.anticheat.data.PlayerData;
import org.hostile.anticheat.event.PositionUpdateEvent;
import org.hostile.anticheat.util.Collisions;
import org.hostile.anticheat.util.PotionEffectType;

@CheckMetadata(type = "Speed", name = "A")
public class SpeedA extends PositionUpdateCheck {

    private static final double AIR_FRICTION = 0.91F;
    private static final double AIR_MOVE_SPEED = 0.026D;

    private static final double SPRINTING_MODIFIER = 1.3D;
    private static final double LAND_MOVEMENT_FACTOR = .16277136F;

    private static final double WATER_MOVEMENT_MODIFIER = 0.800000011920929D;
    private static final double LAVA_MOVEMENT_MODIFIER = 0.5D;

    private static final double JUMP_BOOST = 0.2D;

    private static final double SPEED_THRESHOLD = 1D;

    private static final double SPEED_MODIFIER = 0.2D;
    private static final double SLOWNESS_MODIFIER = 0.15D;

    private static final double ATTACK_MODIFIER = 0.6D;

    private double lastOffsetXZ;
    private double lastFriction = 0.91F;

    public SpeedA(PlayerData playerData) {
        super(playerData);
    }

    @Override
    public void handle(PositionUpdateEvent event) {
        double offsetXZ = event.getOffsetXZ();
        double offsetY = event.getOffsetY();

        double movementSpeed = movementTracker.getWalkSpeed() / 2D;

        if (collisionTracker.getPreviousCollisions().isOnGround()) {
            movementSpeed *= SPRINTING_MODIFIER;

            movementSpeed *= LAND_MOVEMENT_FACTOR / Math.pow(lastFriction, 3);

            if (offsetY > 0.01) { // this can be abused easily
                movementSpeed += JUMP_BOOST;
            }
        } else {
            movementSpeed = AIR_MOVE_SPEED;
            lastFriction = AIR_FRICTION;
        }

        movementSpeed += (SPEED_MODIFIER * potionTracker.getPotionLevel(PotionEffectType.SPEED));
        movementSpeed -= (SLOWNESS_MODIFIER * potionTracker.getPotionLevel(PotionEffectType.SLOWNESS));

        Collisions collisions = collisionTracker.getCollisions();

        if (collisions.isWater()) {
            movementSpeed *= WATER_MOVEMENT_MODIFIER;
        } else if (collisions.isLava()) {
            movementSpeed *= LAVA_MOVEMENT_MODIFIER;
        }

        if (actionTracker.isAttacking()) {
            movementSpeed *= ATTACK_MODIFIER;
        }

        movementSpeed += movementTracker.getVelocityXZ();

        double ratio = (offsetXZ - lastOffsetXZ) / movementSpeed;
        debug("ratio", ratio);

        if (ratio > SPEED_THRESHOLD) {
            if (incrementBuffer(1) > 3) {
                fail("ratio", ratio, "offsetY", offsetY);
            }
        } else {
            decrementBuffer(0.5);
        }

        lastOffsetXZ = offsetXZ * lastFriction;
        lastFriction = collisionTracker.getPreviousCollisions().getFrictionFactor() * AIR_FRICTION;
    }
}
